pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -499,40 +509,64 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
 func (plugin *kubenetNetworkPlugin) teardown(namespace string, name string, id kubecontainer.ContainerID) error {
 	errList := []error{}
 
+	// get POD network status and its IP addresses so we can use the right handlers for each IP family
+	podNetworkStatus, err := plugin.GetPodNetworkStatus(namespace, name, id)
+	if err != nil {
+		errList = append(errList, err)
+	}
+	iplist := podNetworkStatus.IPs
+
 	// Loopback network deletion failure should not be fatal on teardown
 	if err := plugin.delContainerFromNetwork(plugin.loConfig, "lo", namespace, name, id); err != nil {
 		klog.Warningf("Failed to delete loopback network: %v", err)
+		errList = append(errList, err)
 	}
 
 	// no ip dependent actions
 	if err := plugin.delContainerFromNetwork(plugin.netConfig, network.DefaultInterfaceName, namespace, name, id); err != nil {
+		klog.Warningf("Failed to delete %q network: %v", network.DefaultInterfaceName, err)
 		errList = append(errList, err)
 	}
 
+	// get the list of port mappings
 	portMappings, err := plugin.host.GetPodPortMappings(id.ID)
 	if err != nil {
 		errList = append(errList, err)
-	} else if len(portMappings) > 0 {
-		if err = plugin.hostportManager.Remove(id.ID, &hostport.PodPortMapping{
-			Namespace:    namespace,
-			Name:         name,
-			PortMappings: portMappings,
-			HostNetwork:  false,
-		}); err != nil {
-			errList = append(errList, err)
-		}
 	}
 
-	iplist, exists := plugin.getCachedPodIPs(id)
-	if !exists || len(iplist) == 0 {
+	if len(iplist) == 0 {
 		klog.V(5).Infof("container %s (%s/%s) does not have recorded. ignoring teardown call", id, name, namespace)
 		return nil
```

Comment:
I still don't understand this very well, why we have to return here. Do we have to wait for the `delContainerFromNetwork` calls? 
if affirmative, then why we return nil instead of the errors?